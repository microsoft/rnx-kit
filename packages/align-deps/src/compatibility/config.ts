import type { KitConfig } from "@rnx-kit/config";
import { warn } from "@rnx-kit/console";
import { defaultConfig } from "../config";
import { dropPatchFromVersion, modifyManifest } from "../helpers";
import type { AlignDepsConfig, LegacyCheckConfig, Options } from "../types";

const legacyKeys = [
  "capabilities",
  "customProfiles",
  "reactNativeDevVersion",
  "reactNativeVersion",
] as const;

function findLegacyConfigKeys(config: KitConfig): (keyof KitConfig)[] {
  return legacyKeys.filter((key) => key in config);
}

/**
 * Transforms the old config schema into the new one.
 *
 * Note that this config is presented to the user and should therefore be
 * "pretty".
 *
 * @param oldConfig Config in old schema
 * @returns Config in new schema
 */
export function transformConfig({
  capabilities,
  customProfiles,
  kitType,
  manifest,
  reactNativeDevVersion,
  reactNativeVersion,
}: LegacyCheckConfig): AlignDepsConfig {
  const devVersion = dropPatchFromVersion(
    reactNativeDevVersion || reactNativeVersion
  );
  const prodVersion = dropPatchFromVersion(reactNativeVersion);
  return {
    kitType,
    alignDeps: {
      presets: customProfiles
        ? [...defaultConfig.presets, customProfiles]
        : defaultConfig.presets,
      requirements:
        kitType === "app"
          ? [`react-native@${prodVersion}`]
          : {
              development: [`react-native@${devVersion}`],
              production: [`react-native@${prodVersion}`],
            },
      capabilities,
    },
    manifest,
  };
}

/**
 * Migrates the old config schema into the new one, if necessary.
 *
 * This will function will allow users to let `align-deps` update their config.
 * Otherwise, it will tell them how to update their config manually. It will
 * also warn about old config keys that are no longer used.
 *
 * @param config Configuration in the package manifest
 * @param manifestPath Path to the package manifest to check
 * @param options Command line options
 * @returns The config in the new schema
 */
export function migrateConfig(
  config: AlignDepsConfig | LegacyCheckConfig,
  manifestPath: string,
  { migrateConfig }: Options
): AlignDepsConfig {
  if ("alignDeps" in config) {
    const oldKeys = findLegacyConfigKeys(config);
    if (oldKeys.length > 0) {
      const unsupportedKeys = oldKeys
        .map((key) => `'rnx-kit.${key}'`)
        .join(", ");
      warn(
        `${manifestPath}: The following keys are no longer supported: ${unsupportedKeys}`
      );
    }
    return config;
  }

  const newConfig = transformConfig(config);
  const { manifest, ...configOnly } = newConfig;

  if (migrateConfig) {
    const kitConfig = manifest["rnx-kit"];
    if (kitConfig) {
      for (const key of legacyKeys) {
        delete kitConfig[key];
      }
    }

    manifest["rnx-kit"] = {
      ...kitConfig,
      alignDeps: configOnly.alignDeps,
    };

    modifyManifest(manifestPath, manifest);
  } else {
    warn(`${manifestPath}: The config schema has changed. Please update your config to the following:

${JSON.stringify(configOnly, null, 2)}

Or run this command again with '--migrate-config' to update your config automatically.

Support for the old schema will be removed in a future release.`);
  }

  return newConfig;
}
