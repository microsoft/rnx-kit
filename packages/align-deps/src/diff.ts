import { keysOf } from "@rnx-kit/tools-language";
import type { PackageManifest } from "@rnx-kit/tools-node/package";
import type { Changes } from "./types";

export function diff(
  manifest: PackageManifest,
  updatedManifest: PackageManifest
): Changes | undefined {
  const allChanges: Changes = {
    dependencies: [],
    peerDependencies: [],
    devDependencies: [],
  };

  const numChanges = keysOf(allChanges).reduce((count, section) => {
    const changes = allChanges[section];
    const currentDeps = manifest[section] ?? {};
    const updatedDeps = updatedManifest[section] ?? {};

    for (const [dependency, target] of Object.entries(updatedDeps)) {
      const current = currentDeps[dependency];
      if (!current) {
        changes.push({ type: "added", dependency, target });
      } else if (current !== target) {
        changes.push({ type: "changed", dependency, target, current });
      }
    }

    for (const dependency of Object.keys(currentDeps)) {
      if (!updatedDeps[dependency]) {
        changes.push({ type: "removed", dependency });
      }
    }

    return count + changes.length;
  }, 0);

  return numChanges > 0 ? allChanges : undefined;
}

export function stringify(allChanges: Changes, output: string[] = []): string {
  const prefix = "\t  - ";

  for (const [section, changes] of Object.entries(allChanges)) {
    if (changes.length > 0) {
      output.push(`\tIn ${section}:`);
      for (const change of changes) {
        const { type, dependency } = change;
        switch (type) {
          case "added":
            output.push(`${prefix}${dependency} "${change.target}" is missing`);
            break;
          case "changed":
            output.push(
              `${prefix}${dependency} "${change.current}" should be "${change.target}"`
            );
            break;
          case "removed":
            output.push(`${prefix}${dependency} should be removed`);
            break;
        }
      }
    }
  }

  return output.join("\n");
}
