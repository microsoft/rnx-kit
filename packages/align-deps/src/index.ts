export { capabilitiesFor } from "./capabilities";
export { checkPackageManifest } from "./check";
export { cli } from "./cli";
export { updatePackageManifest } from "./manifest";

export type {
  Args,
  CapabilitiesOptions,
  CheckOptions,
  Command,
  DependencyType,
  ExcludedPackage,
  ManifestProfile,
  MetaPackage,
  Package,
  Profile,
  ProfilesInfo,
  ProfileVersion,
  VigilantOptions,
} from "./types";
