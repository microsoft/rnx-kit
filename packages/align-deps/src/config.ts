import type { KitConfig } from "@rnx-kit/config";
import { getKitCapabilities, getKitConfig } from "@rnx-kit/config";
import { error, warn } from "@rnx-kit/console";
import { isPackageManifest, readPackage } from "@rnx-kit/tools-node/package";
import * as path from "path";
import { findBadPackages } from "./findBadPackages";
import type { AlignDepsConfig, ErrorCode, LegacyCheckConfig } from "./types";

type ConfigResult = AlignDepsConfig | LegacyCheckConfig | ErrorCode;

export const defaultConfig: AlignDepsConfig["alignDeps"] = {
  presets: ["microsoft/react-native"],
  requirements: [],
  capabilities: [],
};

export function containsValidPresets(config: KitConfig["alignDeps"]): boolean {
  const presets = config?.presets;
  return !presets || (Array.isArray(presets) && presets.length > 0);
}

export function containsValidRequirements(
  config: KitConfig["alignDeps"]
): boolean {
  const requirements = config?.requirements;
  if (requirements) {
    if (Array.isArray(requirements)) {
      return requirements.length > 0;
    } else if (typeof requirements === "object") {
      return (
        Array.isArray(requirements.production) &&
        requirements.production.length > 0
      );
    }
  }
  return false;
}

/**
 * Loads configuration from the specified package manifest.
 * @param manifestPath The path to the package manifest to load configuration from
 * @returns The configuration; otherwise an error code
 */
export function loadConfig(manifestPath: string): ConfigResult {
  const manifest = readPackage(manifestPath);
  if (!isPackageManifest(manifest)) {
    return "invalid-manifest";
  }

  const badPackages = findBadPackages(manifest);
  if (badPackages) {
    warn(
      `Known bad packages are found in '${manifestPath}':\n` +
        badPackages
          .map((pkg) => `\t${pkg.name}@${pkg.version}: ${pkg.reason}`)
          .join("\n")
    );
  }

  const projectRoot = path.dirname(manifestPath);
  const kitConfig = getKitConfig({ cwd: projectRoot });
  if (!kitConfig) {
    return "not-configured";
  }

  const { kitType = "library", alignDeps, ...config } = kitConfig;
  if (alignDeps) {
    const errors = [];
    if (!containsValidPresets(alignDeps)) {
      errors.push(`${manifestPath}: 'alignDeps.presets' cannot be empty`);
    }
    if (!containsValidRequirements(alignDeps)) {
      errors.push(`${manifestPath}: 'alignDeps.requirements' cannot be empty`);
    }
    if (errors.length > 0) {
      for (const e of errors) {
        error(e);
      }
      return "invalid-configuration";
    }
    return {
      kitType,
      alignDeps: {
        ...defaultConfig,
        ...alignDeps,
      },
      ...config,
      manifest,
    };
  }

  try {
    const {
      capabilities,
      customProfiles,
      reactNativeDevVersion,
      reactNativeVersion,
    } = getKitCapabilities(config);

    return {
      kitType,
      reactNativeVersion,
      ...(config.reactNativeDevVersion ? { reactNativeDevVersion } : undefined),
      capabilities,
      customProfiles,
      manifest,
    };
  } catch (e) {
    return "invalid-configuration";
  }
}
