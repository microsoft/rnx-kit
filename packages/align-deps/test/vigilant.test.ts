import {
  buildManifestProfile,
  checkPackageManifestUnconfigured,
  inspect,
} from "../src/commands/vigilant";
import { defaultConfig } from "../src/config";
import type { AlignDepsConfig } from "../src/types";

jest.mock("fs");

function makeConfig(
  requirements: AlignDepsConfig["alignDeps"]["requirements"],
  manifest: AlignDepsConfig["manifest"] = {
    name: "@rnx-kit/align-deps",
    version: "1.0.0-test",
  }
): AlignDepsConfig {
  return {
    kitType: "library" as const,
    alignDeps: {
      presets: ["microsoft/react-native"],
      requirements,
      capabilities: [],
    },
    manifest,
  };
}

describe("buildManifestProfile()", () => {
  test("builds a package manifest for a single profile version", () => {
    const profile = buildManifestProfile(
      "package.json",
      makeConfig(["react-native@0.70"])
    );
    expect(profile).toMatchSnapshot();
  });

  test("builds a package manifest for multiple profile versions", () => {
    const profile = buildManifestProfile(
      "package.json",
      makeConfig({
        development: ["react-native@0.70"],
        production: ["react-native@0.69 || 0.70"],
      })
    );
    expect(profile).toMatchSnapshot();
  });

  test("includes devOnly packages under `dependencies`", () => {
    const { dependencies, devDependencies, peerDependencies } =
      buildManifestProfile("package.json", makeConfig(["react-native@0.70"]));

    expect("react-native-test-app" in dependencies).toBe(true);
    expect("react-native-test-app" in peerDependencies).toBe(false);
    expect("react-native-test-app" in devDependencies).toBe(true);
  });
});

describe("inspect()", () => {
  const mockManifestProfile = {
    name: "@rnx-kit/align-deps",
    version: "1.0.0",
    dependencies: {
      "react-native": "^0.63.2",
    },
    peerDependencies: {
      "react-native": "^0.63 || ^0.64",
    },
    devDependencies: {
      "react-native": "^0.63.2",
    },
  };

  test("handles empty dependencies", () => {
    const manifest = {
      name: "@rnx-kit/align-deps",
      version: "1.0.0",
    };
    expect(inspect(manifest, mockManifestProfile, false)).toEqual([]);
    expect(inspect(manifest, mockManifestProfile, true)).toEqual([]);
  });

  test("ignores unmanaged dependencies", () => {
    const dependencies = {
      "@babel/core": "^7.0.0",
      "react-native": "0.63.2",
    };
    const manifest = {
      name: "@rnx-kit/align-deps",
      version: "1.0.0",
      dependencies,
    };

    const expectedChanges = [
      {
        name: "react-native",
        from: manifest.dependencies["react-native"],
        to: mockManifestProfile.dependencies["react-native"],
        section: "dependencies",
      },
    ];

    expect(inspect(manifest, mockManifestProfile, false)).toEqual(
      expectedChanges
    );
    expect(manifest.dependencies).toEqual(dependencies);
  });

  test("inspects all dependency types", () => {
    const manifest = {
      name: "@rnx-kit/align-deps",
      version: "1.0.0",
      dependencies: {
        "@babel/core": "^7.0.0",
      },
      peerDependencies: {
        "react-native": "0.63.2",
      },
      devDependencies: {
        "react-native": "0.63.2",
      },
    };

    const expectedChanges = [
      {
        name: "react-native",
        from: manifest.peerDependencies["react-native"],
        to: mockManifestProfile.peerDependencies["react-native"],
        section: "peerDependencies",
      },
      {
        name: "react-native",
        from: manifest.devDependencies["react-native"],
        to: mockManifestProfile.devDependencies["react-native"],
        section: "devDependencies",
      },
    ];

    expect(inspect(manifest, mockManifestProfile, false)).toEqual(
      expectedChanges
    );
  });

  test("modifies the manifest when `write: true`", () => {
    const dependencies = {
      "@babel/core": "^7.0.0",
      "react-native": "0.63.2",
    };
    const manifest = {
      name: "@rnx-kit/align-deps",
      version: "1.0.0",
      dependencies: { ...dependencies },
    };

    const expectedChanges = [
      {
        name: "react-native",
        from: manifest.dependencies["react-native"],
        to: mockManifestProfile.dependencies["react-native"],
        section: "dependencies",
      },
    ];

    expect(inspect(manifest, mockManifestProfile, true)).toEqual(
      expectedChanges
    );
    expect(manifest.dependencies).not.toEqual(dependencies);
  });

  test("does not rewrite peerDependencies if superset", () => {
    const manifest = {
      name: "@rnx-kit/align-deps",
      version: "1.0.0",
      dependencies: {
        react: "^16.8.1",
      },
      peerDependencies: {
        metro: "*",
        react: ">=16.8.0 <18.0.0",
        "react-native": ">=0.64",
      },
      devDependencies: {},
    };
    const profile = {
      name: "@rnx-kit/align-deps",
      version: "1.0.0",
      dependencies: {
        react: "~17.0.1",
      },
      peerDependencies: {
        metro: "^0.66.2",
        react: "~17.0.1",
        "react-native": "^0.66.0-0",
      },
      devDependencies: {},
    };
    const expectedChanges = [
      {
        name: "react",
        from: manifest.dependencies["react"],
        to: profile.dependencies["react"],
        section: "dependencies",
      },
    ];

    expect(inspect(manifest, profile, false)).toEqual(expectedChanges);
  });
});

describe("checkPackageManifestUnconfigured()", () => {
  const rnxKitConfig = require("@rnx-kit/config");
  const fs = require("fs");

  const consoleErrorSpy = jest.spyOn(global.console, "error");

  const defaultOptions = {
    presets: defaultConfig.presets,
    loose: false,
    migrateConfig: false,
    verbose: false,
    write: false,
  };

  beforeEach(() => {
    consoleErrorSpy.mockReset();
    fs.__setMockContent({});
    fs.__setMockFileWriter(() => {
      throw new Error("mock for fs.writeFileSync is not set");
    });
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  test("returns exit code 0 when there are no violations", () => {
    let didWrite = false;
    fs.__setMockFileWriter(() => {
      didWrite = true;
    });

    const result = checkPackageManifestUnconfigured(
      "package.json",
      defaultOptions,
      makeConfig(["react-native@0.70"], {
        name: "@rnx-kit/align-deps",
        version: "1.0.0",
        dependencies: {
          "react-native": "^0.70.0",
        },
      })
    );
    expect(result).toBe("success");
    expect(didWrite).toBe(false);
    expect(consoleErrorSpy).not.toBeCalled();
  });

  test("returns non-zero exit code when there are violations", () => {
    let didWrite = false;
    fs.__setMockFileWriter(() => {
      didWrite = true;
    });

    const result = checkPackageManifestUnconfigured(
      "package.json",
      defaultOptions,
      makeConfig(["react-native@0.70"], {
        name: "@rnx-kit/align-deps",
        version: "1.0.0",
        dependencies: {
          "react-native": "1000.0.0",
        },
      })
    );
    expect(result).not.toBe("success");
    expect(didWrite).toBe(false);
    expect(consoleErrorSpy).toBeCalledTimes(1);
  });

  test("returns exit code 0 when writing", () => {
    let didWrite = false;
    fs.__setMockFileWriter(() => {
      didWrite = true;
    });

    const result = checkPackageManifestUnconfigured(
      "package.json",
      { ...defaultOptions, write: true },
      makeConfig(["react-native@0.70"], {
        name: "@rnx-kit/align-deps",
        version: "1.0.0",
        dependencies: {
          "react-native": "1000.0.0",
        },
      })
    );
    expect(result).toBe("success");
    expect(didWrite).toBe(true);
    expect(consoleErrorSpy).not.toBeCalled();
  });

  test("excludes specified packages", () => {
    let didWrite = false;
    fs.__setMockFileWriter(() => {
      didWrite = true;
    });

    const result = checkPackageManifestUnconfigured(
      "package.json",
      { ...defaultOptions, excludePackages: ["@rnx-kit/align-deps"] },
      makeConfig(["react-native@0.70"], {
        name: "@rnx-kit/align-deps",
        version: "1.0.0",
        dependencies: {
          "react-native": "1000.0.0",
        },
      })
    );
    expect(result).toBe("success");
    expect(didWrite).toBe(false);
    expect(consoleErrorSpy).not.toBeCalled();
  });

  test("uses package-specific custom profiles", () => {
    const fixture = `${__dirname}/__fixtures__/config-custom-profiles-only`;
    const kitConfig = {
      kitType: "library" as const,
      alignDeps: {
        presets: [
          "microsoft/react-native",
          `${fixture}/packageSpecificProfiles.js`,
        ],
        requirements: {
          development: ["react-native@0.64"],
          production: ["react-native@0.64 || 0.65"],
        },
        capabilities: [],
      },
    };
    const inputManifest = {
      name: "@rnx-kit/align-deps",
      version: "1.0.0",
      peerDependencies: {
        react: "17.0.1",
        "react-native": "0.64.0",
      },
      devDependencies: {
        react: "17.0.1",
        "react-native": "0.64.0",
      },
      "rnx-kit": kitConfig,
    };

    rnxKitConfig.__setMockConfig(kitConfig);

    let manifest = undefined;
    fs.__setMockFileWriter((_, content) => {
      manifest = JSON.parse(content);
    });

    const result = checkPackageManifestUnconfigured(
      "package.json",
      { ...defaultOptions, write: true },
      { ...kitConfig, manifest: inputManifest }
    );

    expect(result).toBe("success");
    expect(consoleErrorSpy).not.toBeCalled();
    expect(manifest).toEqual({
      ...inputManifest,
      devDependencies: {
        react: "17.0.2",
        "react-native": "0.64.3",
      },
      peerDependencies: {
        react: "17.0.2",
        "react-native": "0.64.3 || 0.65.2",
      },
    });
  });
});
