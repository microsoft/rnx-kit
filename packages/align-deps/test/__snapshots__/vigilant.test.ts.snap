// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildManifestProfile() builds a package manifest for a single profile version 1`] = `
{
  "dependencies": {
    "@react-native-async-storage/async-storage": "^1.17.10",
    "@react-native-clipboard/clipboard": "^1.10.0",
    "@react-native-community/checkbox": "^0.5.8",
    "@react-native-community/cli": "^9.0.0",
    "@react-native-community/cli-platform-android": "^9.0.0",
    "@react-native-community/cli-platform-ios": "^9.0.0",
    "@react-native-community/datetimepicker": "^6.3.3",
    "@react-native-community/hooks": "^2.8.0",
    "@react-native-community/netinfo": "^9.0.0",
    "@react-native-masked-view/masked-view": "^0.2.7",
    "@react-navigation/native": "^6.0.8",
    "@react-navigation/stack": "^6.2.0",
    "@rnx-kit/react-native-lazy-index": "^2.1.7",
    "jest": "^26.6.3",
    "metro": "^0.72.1",
    "metro-config": "^0.72.1",
    "metro-core": "^0.72.1",
    "metro-react-native-babel-preset": "^0.72.1",
    "metro-react-native-babel-transformer": "^0.72.1",
    "metro-resolver": "^0.72.1",
    "metro-runtime": "^0.72.1",
    "react": "18.1.0",
    "react-dom": "^18.1.0",
    "react-native": "^0.70.0",
    "react-native-base64": "^0.2.1",
    "react-native-floating-action": "^1.22.0",
    "react-native-fs": "^2.18.0",
    "react-native-gesture-handler": "^2.6.0",
    "react-native-macos": "^0.70.0",
    "react-native-modal": "^13.0.0",
    "react-native-popover-view": "^5.0.0",
    "react-native-reanimated": "^2.10.0",
    "react-native-render-html": "^6.1.0",
    "react-native-safe-area-context": "^4.4.1",
    "react-native-screens": "^3.18.2",
    "react-native-shimmer": "^0.5.0",
    "react-native-sqlite-storage": "^6.0.1",
    "react-native-svg": "^12.3.0",
    "react-native-test-app": "^1.6.9",
    "react-native-webview": "^11.23.0",
    "react-native-windows": "^0.70.0",
    "react-test-renderer": "18.1.0",
  },
  "devDependencies": {
    "@react-native-async-storage/async-storage": "^1.17.10",
    "@react-native-clipboard/clipboard": "^1.10.0",
    "@react-native-community/checkbox": "^0.5.8",
    "@react-native-community/cli": "^9.0.0",
    "@react-native-community/cli-platform-android": "^9.0.0",
    "@react-native-community/cli-platform-ios": "^9.0.0",
    "@react-native-community/datetimepicker": "^6.3.3",
    "@react-native-community/hooks": "^2.8.0",
    "@react-native-community/netinfo": "^9.0.0",
    "@react-native-masked-view/masked-view": "^0.2.7",
    "@react-navigation/native": "^6.0.8",
    "@react-navigation/stack": "^6.2.0",
    "@rnx-kit/react-native-lazy-index": "^2.1.7",
    "jest": "^26.6.3",
    "metro": "^0.72.1",
    "metro-config": "^0.72.1",
    "metro-core": "^0.72.1",
    "metro-react-native-babel-preset": "^0.72.1",
    "metro-react-native-babel-transformer": "^0.72.1",
    "metro-resolver": "^0.72.1",
    "metro-runtime": "^0.72.1",
    "react": "18.1.0",
    "react-dom": "^18.1.0",
    "react-native": "^0.70.0",
    "react-native-base64": "^0.2.1",
    "react-native-floating-action": "^1.22.0",
    "react-native-fs": "^2.18.0",
    "react-native-gesture-handler": "^2.6.0",
    "react-native-macos": "^0.70.0",
    "react-native-modal": "^13.0.0",
    "react-native-popover-view": "^5.0.0",
    "react-native-reanimated": "^2.10.0",
    "react-native-render-html": "^6.1.0",
    "react-native-safe-area-context": "^4.4.1",
    "react-native-screens": "^3.18.2",
    "react-native-shimmer": "^0.5.0",
    "react-native-sqlite-storage": "^6.0.1",
    "react-native-svg": "^12.3.0",
    "react-native-test-app": "^1.6.9",
    "react-native-webview": "^11.23.0",
    "react-native-windows": "^0.70.0",
    "react-test-renderer": "18.1.0",
  },
  "peerDependencies": {
    "@react-native-async-storage/async-storage": "^1.17.10",
    "@react-native-clipboard/clipboard": "^1.10.0",
    "@react-native-community/checkbox": "^0.5.8",
    "@react-native-community/datetimepicker": "^6.3.3",
    "@react-native-community/hooks": "^2.8.0",
    "@react-native-community/netinfo": "^9.0.0",
    "@react-native-masked-view/masked-view": "^0.2.7",
    "@react-navigation/native": "^6.0.8",
    "@react-navigation/stack": "^6.2.0",
    "@rnx-kit/react-native-lazy-index": "^2.1.7",
    "react": "18.1.0",
    "react-dom": "^18.1.0",
    "react-native": "^0.70.0",
    "react-native-base64": "^0.2.1",
    "react-native-floating-action": "^1.22.0",
    "react-native-fs": "^2.18.0",
    "react-native-gesture-handler": "^2.6.0",
    "react-native-macos": "^0.70.0",
    "react-native-modal": "^13.0.0",
    "react-native-popover-view": "^5.0.0",
    "react-native-reanimated": "^2.10.0",
    "react-native-render-html": "^6.1.0",
    "react-native-safe-area-context": "^4.4.1",
    "react-native-screens": "^3.18.2",
    "react-native-shimmer": "^0.5.0",
    "react-native-sqlite-storage": "^6.0.1",
    "react-native-svg": "^12.3.0",
    "react-native-webview": "^11.23.0",
    "react-native-windows": "^0.70.0",
  },
  "unmanagedCapabilities": {
    "@react-native-async-storage/async-storage": "storage",
    "@react-native-clipboard/clipboard": "clipboard",
    "@react-native-community/checkbox": "checkbox",
    "@react-native-community/cli": "community/cli",
    "@react-native-community/cli-platform-android": "community/cli-android",
    "@react-native-community/cli-platform-ios": "community/cli-ios",
    "@react-native-community/datetimepicker": "datetime-picker",
    "@react-native-community/hooks": "hooks",
    "@react-native-community/netinfo": "netinfo",
    "@react-native-masked-view/masked-view": "masked-view",
    "@react-navigation/native": "navigation/native",
    "@react-navigation/stack": "navigation/stack",
    "@rnx-kit/react-native-lazy-index": "lazy-index",
    "jest": "jest",
    "metro": "metro",
    "metro-config": "metro-config",
    "metro-core": "metro-core",
    "metro-react-native-babel-preset": "babel-preset-react-native",
    "metro-react-native-babel-transformer": "metro-react-native-babel-transformer",
    "metro-resolver": "metro-resolver",
    "metro-runtime": "metro-runtime",
    "react": "react",
    "react-dom": "react-dom",
    "react-native": "core",
    "react-native-base64": "base64",
    "react-native-floating-action": "floating-action",
    "react-native-fs": "filesystem",
    "react-native-gesture-handler": "gestures",
    "react-native-macos": "core-macos",
    "react-native-modal": "modal",
    "react-native-popover-view": "popover",
    "react-native-reanimated": "animation",
    "react-native-render-html": "html",
    "react-native-safe-area-context": "safe-area",
    "react-native-screens": "screens",
    "react-native-shimmer": "shimmer",
    "react-native-sqlite-storage": "sqlite",
    "react-native-svg": "svg",
    "react-native-test-app": "test-app",
    "react-native-webview": "webview",
    "react-native-windows": "core-windows",
    "react-test-renderer": "react-test-renderer",
  },
}
`;

exports[`buildManifestProfile() builds a package manifest for multiple profile versions 1`] = `
{
  "dependencies": {
    "@react-native-async-storage/async-storage": "^1.17.10",
    "@react-native-clipboard/clipboard": "^1.10.0",
    "@react-native-community/checkbox": "^0.5.8",
    "@react-native-community/cli": "^9.0.0",
    "@react-native-community/cli-platform-android": "^9.0.0",
    "@react-native-community/cli-platform-ios": "^9.0.0",
    "@react-native-community/datetimepicker": "^6.3.3",
    "@react-native-community/hooks": "^2.8.0",
    "@react-native-community/netinfo": "^9.0.0",
    "@react-native-masked-view/masked-view": "^0.2.7",
    "@react-navigation/native": "^6.0.8",
    "@react-navigation/stack": "^6.2.0",
    "@rnx-kit/react-native-lazy-index": "^2.1.7",
    "jest": "^26.6.3",
    "metro": "^0.72.1",
    "metro-config": "^0.72.1",
    "metro-core": "^0.72.1",
    "metro-react-native-babel-preset": "^0.72.1",
    "metro-react-native-babel-transformer": "^0.72.1",
    "metro-resolver": "^0.72.1",
    "metro-runtime": "^0.72.1",
    "react": "18.1.0",
    "react-dom": "^18.1.0",
    "react-native": "^0.70.0",
    "react-native-base64": "^0.2.1",
    "react-native-floating-action": "^1.22.0",
    "react-native-fs": "^2.18.0",
    "react-native-gesture-handler": "^2.6.0",
    "react-native-macos": "^0.70.0",
    "react-native-modal": "^13.0.0",
    "react-native-popover-view": "^5.0.0",
    "react-native-reanimated": "^2.10.0",
    "react-native-render-html": "^6.1.0",
    "react-native-safe-area-context": "^4.4.1",
    "react-native-screens": "^3.18.2",
    "react-native-shimmer": "^0.5.0",
    "react-native-sqlite-storage": "^6.0.1",
    "react-native-svg": "^12.3.0",
    "react-native-test-app": "^1.6.9",
    "react-native-webview": "^11.23.0",
    "react-native-windows": "^0.70.0",
    "react-test-renderer": "18.1.0",
  },
  "devDependencies": {
    "@react-native-async-storage/async-storage": "^1.17.10",
    "@react-native-clipboard/clipboard": "^1.10.0",
    "@react-native-community/checkbox": "^0.5.8",
    "@react-native-community/cli": "^9.0.0",
    "@react-native-community/cli-platform-android": "^9.0.0",
    "@react-native-community/cli-platform-ios": "^9.0.0",
    "@react-native-community/datetimepicker": "^6.3.3",
    "@react-native-community/hooks": "^2.8.0",
    "@react-native-community/netinfo": "^9.0.0",
    "@react-native-masked-view/masked-view": "^0.2.7",
    "@react-navigation/native": "^6.0.8",
    "@react-navigation/stack": "^6.2.0",
    "@rnx-kit/react-native-lazy-index": "^2.1.7",
    "jest": "^26.6.3",
    "metro": "^0.72.1",
    "metro-config": "^0.72.1",
    "metro-core": "^0.72.1",
    "metro-react-native-babel-preset": "^0.72.1",
    "metro-react-native-babel-transformer": "^0.72.1",
    "metro-resolver": "^0.72.1",
    "metro-runtime": "^0.72.1",
    "react": "18.1.0",
    "react-dom": "^18.1.0",
    "react-native": "^0.70.0",
    "react-native-base64": "^0.2.1",
    "react-native-floating-action": "^1.22.0",
    "react-native-fs": "^2.18.0",
    "react-native-gesture-handler": "^2.6.0",
    "react-native-macos": "^0.70.0",
    "react-native-modal": "^13.0.0",
    "react-native-popover-view": "^5.0.0",
    "react-native-reanimated": "^2.10.0",
    "react-native-render-html": "^6.1.0",
    "react-native-safe-area-context": "^4.4.1",
    "react-native-screens": "^3.18.2",
    "react-native-shimmer": "^0.5.0",
    "react-native-sqlite-storage": "^6.0.1",
    "react-native-svg": "^12.3.0",
    "react-native-test-app": "^1.6.9",
    "react-native-webview": "^11.23.0",
    "react-native-windows": "^0.70.0",
    "react-test-renderer": "18.1.0",
  },
  "peerDependencies": {
    "@react-native-async-storage/async-storage": "^1.17.7 || ^1.17.10",
    "@react-native-clipboard/clipboard": "^1.10.0",
    "@react-native-community/checkbox": "^0.5.8",
    "@react-native-community/datetimepicker": "^6.0.2 || ^6.3.3",
    "@react-native-community/hooks": "^2.8.0",
    "@react-native-community/netinfo": "^8.0.0 || ^9.0.0",
    "@react-native-masked-view/masked-view": "^0.2.7",
    "@react-navigation/native": "^6.0.8",
    "@react-navigation/stack": "^6.2.0",
    "@rnx-kit/react-native-lazy-index": "^2.1.7",
    "hermes-engine": "~0.11.0",
    "react": "18.0.0 || 18.1.0",
    "react-dom": "^18.0.0 || ^18.1.0",
    "react-native": "^0.69.0 || ^0.70.0",
    "react-native-base64": "^0.2.1",
    "react-native-floating-action": "^1.22.0",
    "react-native-fs": "^2.18.0",
    "react-native-gesture-handler": "^2.5.0 || ^2.6.0",
    "react-native-macos": "^0.69.0 || ^0.70.0",
    "react-native-modal": "^13.0.0",
    "react-native-popover-view": "^5.0.0",
    "react-native-reanimated": "^2.9.0 || ^2.10.0",
    "react-native-render-html": "^6.1.0",
    "react-native-safe-area-context": "^4.3.1 || ^4.4.1",
    "react-native-screens": "^3.14.1 || ^3.18.2",
    "react-native-shimmer": "^0.5.0",
    "react-native-sqlite-storage": "^6.0.1",
    "react-native-svg": "^12.3.0",
    "react-native-webview": "^11.23.0",
    "react-native-windows": "^0.69.0 || ^0.70.0",
  },
  "unmanagedCapabilities": {
    "@react-native-async-storage/async-storage": "storage",
    "@react-native-clipboard/clipboard": "clipboard",
    "@react-native-community/checkbox": "checkbox",
    "@react-native-community/cli": "community/cli",
    "@react-native-community/cli-platform-android": "community/cli-android",
    "@react-native-community/cli-platform-ios": "community/cli-ios",
    "@react-native-community/datetimepicker": "datetime-picker",
    "@react-native-community/hooks": "hooks",
    "@react-native-community/netinfo": "netinfo",
    "@react-native-masked-view/masked-view": "masked-view",
    "@react-navigation/native": "navigation/native",
    "@react-navigation/stack": "navigation/stack",
    "@rnx-kit/react-native-lazy-index": "lazy-index",
    "jest": "jest",
    "metro": "metro",
    "metro-config": "metro-config",
    "metro-core": "metro-core",
    "metro-react-native-babel-preset": "babel-preset-react-native",
    "metro-react-native-babel-transformer": "metro-react-native-babel-transformer",
    "metro-resolver": "metro-resolver",
    "metro-runtime": "metro-runtime",
    "react": "react",
    "react-dom": "react-dom",
    "react-native": "core",
    "react-native-base64": "base64",
    "react-native-floating-action": "floating-action",
    "react-native-fs": "filesystem",
    "react-native-gesture-handler": "gestures",
    "react-native-macos": "core-macos",
    "react-native-modal": "modal",
    "react-native-popover-view": "popover",
    "react-native-reanimated": "animation",
    "react-native-render-html": "html",
    "react-native-safe-area-context": "safe-area",
    "react-native-screens": "screens",
    "react-native-shimmer": "shimmer",
    "react-native-sqlite-storage": "sqlite",
    "react-native-svg": "svg",
    "react-native-test-app": "test-app",
    "react-native-webview": "webview",
    "react-native-windows": "core-windows",
    "react-test-renderer": "react-test-renderer",
  },
}
`;
